language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the (acyclic) borrow graph for the type and memory safety analysis.
+use crate::nonce::Nonce;
+use mirai_annotations::checked_verify;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// There are two types of edges in the borrow graph.
+/// A strong edge carries more information than a weak edge.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+/// A borrow graph is represented as a map from a source nonce to the set of all edges
+/// coming out of it.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph<T>(BTreeMap<Nonce, BTreeSet<Edge<T>>>);
+
+impl<T> BorrowGraph<T>
+where
+    T: Ord,
+    T: Clone,
+{
+    /// creates a new empty borrow graph
+    pub fn new() -> Self {
+        BorrowGraph(BTreeMap::new())
+    }
+
+    /// adds a fresh nonce
+    pub fn add_nonce(&mut self, nonce: Nonce) {
+        checked_verify!(!self.0.contains_key(&nonce));
```

Comment:
precondition.

Also, instead of requiring the caller to come up with a fresh nonce, why not just allocate a fresh one here?

Reply:
Currently, the type checker decouples the creation of nonces from the representation of borrow relationships among.  The creation is controlled by the code that propagates abstract state across basic blocks.  Certain nonce values are special because they are used to create a canonical representation of an abstract state.  These special nonces and what they mean are known only to the propagation code.  The borrow graph itself does not care about them and only treats nonces as nodes in the graph.